// @passoa/cxb/lib/downloader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
var cmLog_1 = require("./cmLog");
var fs = require("fs-extra");
var path = require("path");
var axios_1 = require("axios");
var compressing_1 = require("compressing");
var Downloader = /** @class */ (function () {
    function Downloader(options) {
        this.done = 0;
        this.length = 0;
        this.unzipTotal = 0;
        this.unzipNo = 0;
        this.notifyOk = function (result) { };
    }
    Downloader.prototype.downloadAll = function (task) {
        return __awaiter(this, void 0, void 0, function () {
            var op, task_1, task_1_1, iter;
            var e_1, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        op = [];
                        try {
                            for (task_1 = __values(task), task_1_1 = task_1.next(); !task_1_1.done; task_1_1 = task_1.next()) {
                                iter = task_1_1.value;
                                op.push(this.downloadFile(iter));
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (task_1_1 && !task_1_1.done && (_a = task_1.return)) _a.call(task_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        return [4 /*yield*/, Promise.all(op)];
                    case 1:
                        _b.sent();
                        console.log('\ndownload all end\n');
                        return [2 /*return*/];
                }
            });
        });
    };
    Downloader.prototype.downloadFile = function (iter) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            axios_1.default({
                method: 'get',
                url: iter.src,
                responseType: 'stream'
            })
                .then(function (response) {
                _this.length += parseInt(response.headers['content-length'], 10);
                fs.ensureDirSync(path.dirname(iter.dst));
                response.data.on('data', function (chunk) {
                    _this.done += chunk.length;
                    if (isNaN(_this.length)) {
                        cmLog_1.slog("Downloading " + (_this.done / 1024).toFixed(2) + "/ unknow kb");
                    }
                    else {
                        cmLog_1.slog("Downloading " + (_this.done / 1024).toFixed(2) + "/ " + (_this.length / 1024).toFixed(2) + " kb");
                    }
                });
                fs.ensureFileSync(iter.dst);
                response.data.pipe(fs
                    .createWriteStream(iter.dst)
                    .on('close', function () {
                    resolve(true);
                })
                    .on('error', function (err) {
                    throw new Error("some error was happend with write file " + iter.src);
                })
                    .on('finish', function () { }));
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Downloader.prototype.getExt = function (filename, extlist) {
        var e_2, _a;
        var found = '';
        try {
            for (var extlist_1 = __values(extlist), extlist_1_1 = extlist_1.next(); !extlist_1_1.done; extlist_1_1 = extlist_1.next()) {
                var ext = extlist_1_1.value;
                if (filename.endsWith('.' + ext)) {
                    if (found.length < ext.length) {
                        found = ext;
                    }
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (extlist_1_1 && !extlist_1_1.done && (_a = extlist_1.return)) _a.call(extlist_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return found;
    };
    Downloader.prototype.unzipAll = function (task) {
        return __awaiter(this, void 0, void 0, function () {
            var op, task_2, task_2_1, iter, ext;
            var e_3, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        op = [];
                        try {
                            for (task_2 = __values(task), task_2_1 = task_2.next(); !task_2_1.done; task_2_1 = task_2.next()) {
                                iter = task_2_1.value;
                                ext = this.getExt(iter.src, ['tgz', 'tar.gz', 'zip', 'gz', 'gzip']);
                                switch (ext) {
                                    case 'tgz':
                                    case 'tar.gz':
                                        op.push(this.uncompressingTgz(iter));
                                        break;
                                    case 'zip':
                                        op.push(this.uncompressingZip(iter));
                                        break;
                                    case 'gz':
                                    case 'gzip':
                                        op.push(this.uncompressingGZip(iter));
                                        break;
                                }
                            }
                        }
                        catch (e_3_1) { e_3 = { error: e_3_1 }; }
                        finally {
                            try {
                                if (task_2_1 && !task_2_1.done && (_a = task_2.return)) _a.call(task_2);
                            }
                            finally { if (e_3) throw e_3.error; }
                        }
                        return [4 /*yield*/, Promise.all(op)];
                    case 1:
                        _b.sent();
                        console.log('\nunzip all end\n');
                        return [2 /*return*/];
                }
            });
        });
    };
    Downloader.prototype.strip = function (str, deep) {
        var arr = path.normalize(str).split(path.sep);
        arr.splice(0, arr.length < deep ? arr.length - 1 : deep);
        return arr.join(path.sep);
    };
    Downloader.prototype.onEntry = function (iter, header, stream, next) {
        var _this = this;
        stream.on('end', next);
        if (iter.option && iter.option.strip)
            header.name = this.strip(header.name, iter.option.strip);
        if (header.type === 'file') {
            this.unzipTotal++;
            fs.ensureDirSync(path.dirname(path.join(iter.dst, header.name)));
            stream.pipe(fs.createWriteStream(path.join(iter.dst, header.name)).on('close', function () {
                _this.unzipNo++;
                _this.notifyFinish();
            }));
        }
        else {
            // directory
            fs.ensureDirSync(path.dirname(path.join(iter.dst, header.name)));
            stream.resume();
        }
    };
    Downloader.prototype.notifyFinish = function () {
        if (this.notifyOk && this.unzipNo >= this.unzipTotal)
            this.notifyOk(true);
    };
    Downloader.prototype.uncompressingTgz = function (iter) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            new compressing_1.tgz.UncompressStream({ source: iter.src })
                .on('error', function (err) {
                reject(err);
            })
                .on('finish', function () {
                _this.notifyOk = resolve;
                _this.notifyFinish();
            }) // uncompressing is done
                .on('entry', function (header, stream, next) {
                _this.onEntry(iter, header, stream, next);
            });
        });
    };
    Downloader.prototype.uncompressingZip = function (iter) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            new compressing_1.zip.UncompressStream({ source: iter.src })
                .on('error', function (err) {
                reject(err);
            })
                .on('finish', function () {
                _this.notifyOk = resolve;
                _this.notifyFinish();
            }) // uncompressing is done
                .on('entry', function (header, stream, next) {
                _this.onEntry(iter, header, stream, next);
            });
        });
    };
    Downloader.prototype.uncompressingGZip = function (iter) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            new compressing_1.gzip.UncompressStream({ source: iter.src })
                .on('error', function (err) {
                reject(err);
            })
                .on('finish', function () {
                _this.notifyOk = resolve;
                _this.notifyFinish();
            }) // uncompressing is done
                .on('entry', function (header, stream, next) {
                _this.onEntry(iter, header, stream, next);
            });
        });
    };
    return Downloader;
}());
exports.Downloader = Downloader;
