// @passoa/cxb/lib/uploader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_extra_1 = require("fs-extra");
var cmLog_1 = require("./cmLog");
var FormData = require("form-data");
var axios_1 = require("axios");
var compressing_1 = require("compressing");
var stream_1 = require("stream");
var fs_1 = require("fs");
var Uploader = /** @class */ (function () {
    function Uploader() {
    }
    Uploader.prototype.packTgz = function (src, dst) {
        var _this = this;
        return new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
            var ts, destStream, len;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ts = new compressing_1.tgz.Stream();
                        ts.addEntry(src);
                        return [4 /*yield*/, fs_extra_1.ensureFile(dst)];
                    case 1:
                        _a.sent();
                        destStream = fs_extra_1.createWriteStream(dst);
                        len = 0;
                        ts.on('data', function (data) {
                            cmLog_1.slog("ziped :" + (len += data.length));
                        });
                        stream_1.pipeline(ts, destStream, function (ev) {
                            if (ev) {
                                console.log(ev);
                                reject(-1);
                            }
                            else {
                                resolve(0);
                            }
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    };
    Uploader.prototype.readFileByPath = function (path) { };
    Uploader.prototype.upload = function (url, path, token, opt) {
        var _this = this;
        if (opt === void 0) { opt = { method: 'put', form: false }; }
        return new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
            var form, headers;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = {};
                        if (!form) return [3 /*break*/, 1];
                        form = new FormData(); // FormData 对象
                        form.append('files', fs_extra_1.createReadStream(path));
                        headers = form.getHeaders();
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, new Promise(function (resolve, reject) {
                            fs_1.readFile(path, function (err, data) {
                                if (err) {
                                    reject(err);
                                }
                                else {
                                    resolve(data);
                                }
                            });
                        })];
                    case 2:
                        form = _a.sent();
                        headers['Content-Length'] = form.length;
                        _a.label = 3;
                    case 3:
                        headers.Authorization = "Basic " + token;
                        axios_1.default({
                            method: 'put',
                            url: url,
                            data: form,
                            headers: headers,
                            transformRequest: function (data, headers) {
                                // Do whatever you want to transform the data
                                if (data._valuesToMeasure) {
                                    data._valuesToMeasure[0].on('data', function (data) {
                                        console.log('uploaded:' + data.length);
                                    });
                                }
                                return data;
                            }
                        })
                            // axios({
                            // 	method: 'put',
                            // 	url: url,
                            // 	data: form,
                            // 	headers: headers,
                            // 	onUploadProgress: (progressEvent) => {
                            // 		let complete = ((progressEvent.loaded / progressEvent.total * 100) | 0) + '%';
                            // 		console.log('upload:', complete);
                            // 	}
                            // })
                            .then(function (res) {
                            //根据服务器返回进行处理
                            if (res.status > 199 && res.status < 300) {
                                console.log('upload done');
                                resolve(0);
                            }
                            else {
                                console.log('upload fail');
                                reject(-1);
                            }
                        })
                            .catch(function (error) {
                            console.log("[uploaded error]:", error.message);
                            reject(-2);
                        });
                        return [2 /*return*/];
                }
            });
        }); });
    };
    return Uploader;
}());
exports.Uploader = Uploader;
